### PR TITLE
ci/release: Use bazel to build binary bundles

### PR DESCRIPTION
**Pull request by phlax**
- add release bundle targets
- use zstd for bundling release tarball
- cleanup/optimize Docker recipe

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: distribution/binary/compiler.bzl [hunk 1]
```diff
@@ -0,0 +1,47 @@
+def _config_mode_impl(settings, attr):
+    _ignore = settings
+    return {"//command_line_option:compilation_mode": attr.mode}
+
+mode_transition = transition(
```
Comment (phlax) [inline distribution/binary/compiler.bzl:5]:
```suggestion
compilation_mode_transition = transition(
```

File: distribution/binary/compiler.bzl [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            (
+                ctx.attr.name,
+                ctx.var["COMPILATION_MODE"],
+                target.files.to_list()[0].basename,
+            ),
+        )
+        ctx.actions.run(
+            outputs = [output_file],
+            inputs = [target.files.to_list()[0]],
+            arguments = [target.files.to_list()[0].path, output_file.path],
```
Comment (phlax) [inline distribution/binary/compiler.bzl:26]: dont use `to_list()`

File: ci/do_ci.sh [hunk 3]
```diff
@@ -224,14 +224,21 @@ if [[ "$CI_TARGET" == "bazel.release" ]]; then
...
+
+  # Run release tests
   echo "Testing ${TEST_TARGETS[*]} with options: ${BAZEL_BUILD_OPTIONS[*]}"
   bazel_with_collection test "${BAZEL_BUILD_OPTIONS[@]}" --remote_download_minimal -c opt "${TEST_TARGETS[@]}"
 
-  echo "bazel release build..."
-  bazel_envoy_binary_build release
+  # Run release binaries
+  bazel build "${BAZEL_BUILD_OPTIONS[@]}" //distribution/binary:release
+  bazel build "${BAZEL_BUILD_OPTIONS[@]}" -c opt //test/tools/schema_validator:schema_validator_tool
```
Comment (phlax) [inline ci/do_ci.sh:236]: add comment about not doing this

Comment (phlax) [reply ci/do_ci.sh:236]: or move comment here

**Comment by phlax** [comment]
this is slightly slower in the RBE case as it has to zip/collate large files

adding a bazel `pigz` recipe would probably help, and is something i reckon could give us quite a few other speedups - ill experiment

File: ci/Dockerfile-envoy [hunk 4]
```diff
@@ -6,40 +6,52 @@ ARG ENVOY_VRP_BASE_IMAGE=envoy-base
...
+CMD ["envoy", "-c", "/etc/envoy/envoy.yaml"]
+RUN mkdir -p /etc/envoy \
+    && adduser --group --system envoy
+ENTRYPOINT ["/docker-entrypoint.sh"]
 # NB: Adding this here means that following steps, for example updating the system packages, are run
 #   when the version file changes. This should mean that a release version will always update.
 #   In PRs this will just use cached layers unless either this file changes or the version has changed.
 ADD VERSION.txt /etc/envoy
-RUN apt-get -qq update && apt-get -qq upgrade -y \
+RUN --mount=type=tmpfs,target=/var/cache/apt \
```
[the remark this answers is not in the document]
Comment (phlax) [reply ci/Dockerfile-envoy:27]: doesnt seem to

**Comment by phlax** [comment]
> this is slightly slower in the RBE case 

not any longer, this is now significantly faster

File: distribution/binary/BUILD [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+pkg_files(
+    name = "debug_files",
+    srcs = [":envoy-debug"],
+    prefix = "debug",
+)
+
+# This target compresses large files together. For the most part
+# doing this locally will be faster than using RBE, and the files
```
Comment (phlax) [inline distribution/binary/BUILD:39]: rm comment

File: distribution/binary/BUILD [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        "//distribution:envoy-contrib-binary": "dbg/envoy-contrib",
+        "//distribution:envoy-contrib-dwarf": "dbg/envoy-contrib.dwp",
+        "//external:su-exec": "utils/su-exec",
+    },
+)
+
+pkg_files(
+    name = "release_files",
+    srcs = [":envoy"],
+    # prefix = "",
```
Comment (phlax) [inline distribution/binary/BUILD:28]: rm

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Pull request merged**